File: padlock/src/main/java/com/pyamsoft/padlock/main/MainActivity.java
/*
 * Copyright 2017 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.padlock.main;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.view.ViewCompat;
import android.support.v7.app.ActionBar;
import android.support.v7.preference.PreferenceManager;
import com.pyamsoft.padlock.BuildConfig;
import com.pyamsoft.padlock.Injector;
import com.pyamsoft.padlock.R;
import com.pyamsoft.padlock.databinding.ActivityMainBinding;
import com.pyamsoft.padlock.onboard.firstlaunch.OnboardFragment;
import com.pyamsoft.pydroid.ui.about.AboutLibrariesFragment;
import com.pyamsoft.pydroid.ui.sec.TamperActivity;
import com.pyamsoft.pydroid.ui.util.AnimUtil;
import com.pyamsoft.pydroid.util.AppUtil;
import javax.inject.Inject;
import timber.log.Timber;

public class MainActivity extends TamperActivity {

  @SuppressWarnings("WeakerAccess") @Inject MainPresenter presenter;
  @SuppressWarnings("WeakerAccess") ActivityMainBinding binding;

  @Override public void onCreate(final @Nullable Bundle savedInstanceState) {
    setTheme(R.style.Theme_PadLock_Light);
    super.onCreate(savedInstanceState);
    binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
    PreferenceManager.setDefaultValues(getApplicationContext(), R.xml.preferences, false);

    Injector.get().provideComponent().inject(this);

    setAppBarState();
    showOnboardingOrDefault();
  }

  @Override protected void onStop() {
    super.onStop();
    presenter.stop();
  }

  @Override protected void onPause() {
    super.onPause();
    if (isFinishing() || isChangingConfigurations()) {
      Timber.d(
          "Even though a leak is reported, this should dismiss the window, and clear the leak");
      binding.toolbar.getMenu().close();
      binding.toolbar.dismissPopupMenus();
    }
  }

  @NonNull @Override public String provideApplicationName() {
    return "PadLock";
  }

  @Override public int getCurrentApplicationVersion() {
    return BuildConfig.VERSION_CODE;
  }

  private void setAppBarState() {
    setSupportActionBar(binding.toolbar);
    binding.toolbar.setTitle(getString(R.string.app_name));
    ViewCompat.setElevation(binding.toolbar, AppUtil.convertToDP(this, 4));
  }

  @Override protected void onDestroy() {
    super.onDestroy();
    presenter.destroy();
    binding.unbind();
  }

  @Override public void onBackPressed() {
    final FragmentManager fragmentManager = getSupportFragmentManager();
    final int backStackCount = fragmentManager.getBackStackEntryCount();
    if (backStackCount > 0) {
      fragmentManager.popBackStackImmediate();
    } else {
      super.onBackPressed();
    }
  }

  @Override protected void onPostResume() {
    super.onPostResume();
    AnimUtil.animateActionBarToolbar(binding.toolbar);
  }

  @NonNull @Override protected String getSafePackageName() {
    return "com.pyamsoft.padlock";
  }

  @NonNull @Override protected String[] getChangeLogLines() {
    final String line1 = "BUGFIX: Bugfixes and improvements";
    final String line2 = "BUGFIX: Removed all Advertisements";
    final String line3 = "BUGFIX: Faster loading of Open Source Licenses page";
    return new String[] { line1, line2, line3 };
  }

  @NonNull @Override protected String getVersionName() {
    return BuildConfig.VERSION_NAME;
  }

  @Override public int getApplicationIcon() {
    return R.mipmap.ic_launcher;
  }

  /**
   * Called from Onboarding fragments
   */
  public void onOnboardingCompleted() {
    final FragmentManager fragmentManager = getSupportFragmentManager();
    final Fragment onboarding = fragmentManager.findFragmentByTag(OnboardFragment.TAG);
    if (onboarding != null) {
      fragmentManager.beginTransaction().remove(onboarding).commitNow();
    }

    showOnboardingOrDefault();
  }

  private void showOnboardingOrDefault() {
    presenter.showOnboardingOrDefault(new MainPresenter.OnboardingCallback() {
      @Override public void onShowOnboarding() {
        FragmentManager fm = getSupportFragmentManager();
        if (fm.findFragmentByTag(OnboardFragment.TAG) == null) {
          fm.beginTransaction()
              .replace(R.id.fragment_container, new OnboardFragment(), OnboardFragment.TAG)
              .commitNow();
        }
        prepareActivityForOnboarding();
      }

      @Override public void onShowDefaultPage() {
        // Set normal navigation
        FragmentManager fm = getSupportFragmentManager();
        if (fm.findFragmentByTag(OnboardFragment.TAG) != null) {
          Timber.w(
              "Show default page was called, but Onboarding fragment still exists, continue onboarding");
          prepareActivityForOnboarding();
        } else {
          // Un hide the action bar in case it was hidden
          final ActionBar actionBar = getSupportActionBar();
          if (actionBar != null) {
            if (!actionBar.isShowing()) {
              actionBar.show();
            }
          }

          if (fm.findFragmentByTag(MainFragment.TAG) == null
              && fm.findFragmentByTag(AboutLibrariesFragment.TAG) == null) {
            Timber.d("Load default page");
            fm.beginTransaction()
                .replace(R.id.fragment_container, new MainFragment(), MainFragment.TAG)
                .commitNow();
          } else {
            Timber.w("Default page or About libraries was already loaded");
          }
        }
      }
    });
  }

  void prepareActivityForOnboarding() {
    // Hide the action bar
    final ActionBar actionBar = getSupportActionBar();
    if (actionBar != null) {
      if (actionBar.isShowing()) {
        actionBar.hide();
      }
    }
  }
}

File: padlock/src/main/java/com/pyamsoft/padlock/main/MainActivity.kt
/*
 * Copyright (C) 2018 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.padlock.main

import android.content.Intent
import android.os.Bundle
import android.view.View
import com.pyamsoft.padlock.BuildConfig
import com.pyamsoft.padlock.Injector
import com.pyamsoft.padlock.PadLockComponent
import com.pyamsoft.padlock.R
import com.pyamsoft.padlock.helper.ListStateUtil
import com.pyamsoft.padlock.service.ServiceManager
import com.pyamsoft.pydroid.ui.about.AboutFragment
import com.pyamsoft.pydroid.ui.rating.ChangeLogBuilder
import com.pyamsoft.pydroid.ui.rating.RatingActivity
import com.pyamsoft.pydroid.ui.rating.buildChangeLog
import com.pyamsoft.pydroid.ui.theme.Theming
import com.pyamsoft.pydroid.ui.util.commit
import timber.log.Timber
import javax.inject.Inject

class MainActivity : RatingActivity() {

  @field:Inject internal lateinit var mainView: MainView
  @field:Inject internal lateinit var serviceManager: ServiceManager
  @field:Inject internal lateinit var theming: Theming

  override val versionName: String
    get() = BuildConfig.VERSION_NAME

  override val applicationIcon: Int
    get() = R.mipmap.ic_launcher

  override val rootView: View
    get() = mainView.root()

  override val changeLogLines: ChangeLogBuilder
    get() =
      buildChangeLog {
        change("New icon style")
        change("Better open source license viewing experience")
        bugfix("Lots of tiny optimizations for locking applications.")
      }

  override fun onCreate(savedInstanceState: Bundle?) {
    Injector.obtain<PadLockComponent>(applicationContext)
        .plusMainComponent()
        .mainActivity(this)
        .build()
        .inject(this)

    if (theming.isDarkTheme()) {
      setTheme(R.style.Theme_PadLock_Dark_Normal)
    } else {
      setTheme(R.style.Theme_PadLock_Light_Normal)
    }

    super.onCreate(savedInstanceState)
    mainView.create()

    mainView.onToolbarNavigationClicked { onBackPressed() }

    showDefaultPage()
  }

  private fun showDefaultPage() {
    // Set normal navigation
    val fm = supportFragmentManager

    if (fm.findFragmentByTag(MainFragment.TAG) == null && !AboutFragment.isPresent(this)) {
      Timber.d("Load default page")
      fm.beginTransaction()
          .add(R.id.fragment_container, MainFragment(), MainFragment.TAG)
          .commit(this)
    } else {
      Timber.w("Default page or About libraries was already loaded")
    }
  }

  override fun onNewIntent(intent: Intent?) {
    super.onNewIntent(intent)
    setIntent(intent)
  }

  override fun onPause() {
    super.onPause()
    if (isFinishing || isChangingConfigurations) {
      mainView.closeToolbarMenu()
    }
  }

  override fun onDestroy() {
    super.onDestroy()

    if (!isChangingConfigurations) {
      ListStateUtil.clearCache()
    }
  }

  override fun onPostResume() {
    super.onPostResume()
    // Try to start service, will not if we do not have permission
    serviceManager.startService(false)
  }
}
